Add LOCKSS-o-matic support

Hwesta opened the pull request:
Adds support for LOCKSS (via LOCKSS-o-matic using the SWORD2 API) as an AIP Storage destination.  Details found https://www.archivematica.org/wiki/LOCKSS_Integration

In short,
- SS requests a service document from LOM, and parses it for max AU size, and collection to create AIPs in
- If the package is larger than the max AU size, use tar to break the file into chunks
- SS posts an atom entry creation document to LOM, including the URL(s) of the package to be uploaded
- SS parses the returned deposit receipt for the state and edit IRIs
- When package status is updated, query the state IRI for status.  If all servers are in agreement, optionally delete local copy of file

Other features added:
- common NSMAP for use with xml
- storage service version function (common/version.py)
- update_package_status function,which queries the space to check the current status of a package.  For LOM, will resend atom creation and/or query for the server states
- move_from SS but not move_to
- Package.misc_attributes - for storing information about packages that is only needed in some cases, eg state and edit IRI for LOM

Includes code from #5

Hwesta committed:
Added LOCKSS-o-matic stubs

refs #5425

Added a Space class for the LOCKSS-o-matic.  Functionality relating
to LOM on the changes to the METS file are stubs, since those
requirements not yet decided on.

Updated store AIP and download file functions to work with LOM, and
added a stub callback function for when LOM is done storing the file.
Added a new status for a package for when it is on the Storage Server,
but not yet in LOM.

Split chunks of an AIP are found at <aip_uuid>/download/<chunk_number>/

Hwesta committed:
LOCKSS Space can parse service document

Add function to fetch and parse a service document, update the au_size and
collection_iri.

Added dependency on sword2 client library. Github version is newer and more
featureful than the version on PyPI. Install dependency from github.

Hwesta committed:
Add common NSMAP

Add a common namespace map for use when working with METS and XML.
Update existing lxml code to use it. Add ipython as a development
dependency.

Hwesta committed:
Create AtoM entry for AIP

refs #6009

Create the AtoM entry with the title, ID, agent, and description of the
AIP. Title, agent and description parsed from the METS file Dublin Core
title, description and an 'organization' agent.  Extract state IRI for
polling LOCKSS-o-matic from the deposit receipt.

Hwesta committed:
Update LOCKSS-o-matic service doc and resource creation

refs #6009

Update LOCKSS handling of service document and atom entries to create
resources to match API changes.

Add checksumType parsing to service document.  Resource creation uses
lom:content tags, add checksum and checksum type to content tags.

Code moved to LOCKSS Space from Package, since it deals more with
Space-specific problems than Package.

Hwesta committed:
LOCKSS WIP - edit & state IRI

refs #5425

Properly parse the edit and state IRIs from the deposit receipt.  Add sword
and Lockss-o-matic related namespaces to the namespace map.  Start on
polling for state, comments and logs where work needed.

Hwesta committed:
Initial LOCKSS tests

Add pytest and pytest-django as testing dependencies.  Add testing fixtures
- initial data needed for everything, and a lockss-o-matic specific one.
Start on adding tests; currently only work if the modified simple sword
server is running.

Hwesta committed:
Update for LOM changes

refs #5425

Update namespace for LOM, and make common.utils the only place it is
defined.  Add content provider ID, input by the user, and used to identify
the SS to LOM.

Hwesta committed:
File splitting for LOCKSS uses tar

refS #5425

Splitting the file into chunks of no more than au_size uses tar instead of
split.  A tar file has headers and stores some of its own metadata, making
it a safer choice.

Hwesta committed:
Update Lockssomatic to use new infrastructure

refs #5425

Update Lockssomatic to use (post) move to/from storage service.  Removed
now duplicate _store_aip_lom.  Store AIP calls post move functions.

Hwesta committed:
Sizes to/from LOM are in kB

refs #5425

maxUploadSize in the service document, and lom:content size in the atom
entry are both in kilobytes, while all internal sizes are stored in bytes.
Add unit conversion.  Outgoing size rounded up.

Hwesta committed:
Save State & Edit IRI in Packages.misc

refs #5425

Save state & edit IRI from deposit receipt into the misc attributes field
of Packages.

Hwesta committed:
Add package Update Status button for LOCKSS

refs #5425

Add button to check the status of a Package.  Calls update_package_status
in Lockssomatic Space, which queries the state IRI and sets it to success if
all the server states are 'agreement'.

Hwesta committed:
Update pointer file for entry creation

refs #5425

Update structMap and fileGrp, for when a package is split or unsplit.  Add
event and SS agent when the file is split.

Add helper function for writing PREMIS events and agents.  Update structMap
and fileGrp of pointer file with information on chunks.

Add version.py for tracking storage service version and providing string
display of it.

Hwesta committed:
Update pointer file with LOCKSS URLs

refs #5425

Query the state_iri for this package and parse it for the server states. If
all are in agreement, add those URLs to the pointer file for each LOCKSS
chunk.  Works for both split and unsplit packages.

Factor out generation of download URL, as that is the ID of lom:content
elements in the SWORD statement.

Hwesta committed:
Minor fixes

refs #5425

Misc minor fixes: update_package_status idempotent (removes LOCKSS URLs
before adding new ones), download path returns correct URL for tar changes,
checksum name is always uppercase, others.

Hwesta committed:
Update pointer file - delete local copy

refs #5425

Delete LOCKSS chunks once AIP is safely stored, and remove their local URLs
from the pointer file.  If configured, delete local copy of AIP and update
pointer file: add deletion event, remove local URLs, etc.  Ensure that
multiple delete events not added if update status called repeatedly.

Hwesta committed:
Download URL is absolute

refs #5425

Require the externally available URL to be configured for LOM Spaces, as
programatically determining the IP is unreliable.  Download URLs provided
to Lockss-o-matic prepend this to the relative path.

Service Document IRI converted to URLField to ensure it is a valid URL.
Collection IRI left as CharField, as it is read from the SD, and any
validation error would probably get buried.

Hwesta committed:
Update LOM to stop harvesting before deleting files

refs #5425

Require LOM to be updated to stop harvesting before deleting local
copies of files.  If updating the package status fails, return an error
string to display.  Refactor code relating to delete files into its own
function.

Hwesta committed:
Update package status will re-post to LOM

refs #5425

Updating package status will re-submit the atom entry to LOM, if the
first entry failed. post_move_from_storage_service updated to be
idempotent, including all the functions it calls.

Refactored split_package to include all modifications to the pointer
file, and create_resource to only create the atom entry.  split_package
will return the output files if a split has already been performed.

Generate checksum moved to utils.  Parsed pointer file for current
Package made an attribute of Lockssomatic, populated on demand, so it
is easier for functions to reuse it.

Hwesta committed:
LOCKSS AU chunk numbers from 1

refs #5425

Update download_url and download_path to both be indexed from 1,
instead of mismatched.

Hwesta committed:
Delete AIP from LOCKSS

refs #6542

Notify LOM that files will be deleted, but ignore errors.  Delete split
files and original AIP from local filesystem.  _delete_files refactored
to support this.

Hwesta committed:
Update migration

Update migration for LOCKSS-o-matic table and Package.misc_attributes.

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -490,6 +508,497 @@ def post_move_from_storage_service(self, staging_path, destination_path, package
+
+    def move_to_storage_service(self, src_path, dest_path, dest_space):
+        """ Moves src_path to dest_space.staging_path/dest_path. """
+        # Check if in SS internal, if not then fetch from LOM
```
Is this going to be done here?

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -490,6 +508,497 @@ def post_move_from_storage_service(self, staging_path, destination_path, package
+            state_iri = receipt.atom_statement_iri
+            edit_iri = receipt.edit
+        except AttributeError:
+            logging.warning('Unable to contact LOCKSS for package %s', package.uuid)
```
When would this AttributeError occur? If `receipt is None`?

Hwesta replied:
When receipt is None, or if there's any error with receipt that results in atom_state_iri or edit_iri not being populate (eg not in the returned XML, wrong version of the SWORD2 library that doesn't populate them, etc)

mistydemeo replied:
Maybe add a comment to this effect so I remember this, and don't try to figure out why this happening once I forget? :D

Hwesta replied:
Done, and added a check that state_iri and edit_iri are populated before adding them to misc_attributes.

mistydemeo commented:
:thumbsup:

Hwesta committed:
Fix LOCKSS bugs

refs #5425

When updating service document fails, don't continue the function as
LOM is probably not accessible.  To account for earlier versions of AM
pointer file, add fileGrp USE="Archival Information Package", as LOM
processing expects it.

Hwesta commented:
Rebased on current qa/0.x, fixed some rebase errors, and added a small fix for older versions of the pointer file (so this can be released before artefactual/archivematica#28 in 1.2)

mistydemeo commented on storage_service/common/utils.py:
```diff
@@ -100,8 +118,101 @@ def download_file_stream(filepath, temp_dir=None):
+    """
+    try:
+        checksum = hashlib.new(checksum_type)
+    except ValueError:  # Invalid checksum_type
```
Maybe it's worth leaving the raise in this case - silently returning a different checksum type could lead to weird hard-to-debug issues later.

Hwesta replied:
Move `ValueError` checks to the calling function to make this more explicit.

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -368,6 +416,7 @@ def prepend_urls(self):
         return [
             url(r"^(?P<resource_name>%s)/(?P<%s>\w[\w/-]*)/delete_aip%s$" % (self._meta.resource_name, self._meta.detail_uri_name, trailing_slash()), self.wrap_view('delete_aip_request'), name="delete_aip_request"),
             url(r"^(?P<resource_name>%s)/(?P<%s>\w[\w/-]*)/extract_file%s$" % (self._meta.resource_name, self._meta.detail_uri_name, trailing_slash()), self.wrap_view('extract_file_request'), name="extract_file_request"),
+            url(r"^(?P<resource_name>%s)/(?P<%s>\w[\w/-]*)/download/(?P<chunk_number>\d+)%s$" % (self._meta.resource_name, self._meta.detail_uri_name, trailing_slash()), self.wrap_view('download_request'), name="download_lockss"),
```
Not sure how I feel about LOCKS numbered chunk downloads sharing an API endpoint w/ regular downloads; it seems like it's a pretty dissimilar case, unless I'm misunderstanding.

Hwesta replied:
Both are trying to download the entire AIP (or, in the case of chunks, a pre-split part of the AIP).  `Package.get_download_path` will return the package's full path unless both a chunk number is provided and the package is stored in LOM.  Can you clarify what seems dissimilar about the cases?

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -486,6 +501,497 @@ def post_move_from_storage_service(self, staging_path, destination_path, package
+    def post_move_from_storage_service(self, staging_path, destination_path, package):
+        # LOCKSS can only save packages in the storage service, since it needs
+        # to track information on it over time
+        if package is None:
```
In what circumstances is `package` `None`? (I know this isn't specific to LOCKSS.)

Hwesta replied:
Hopefully, never.  `post_move_from_storage_service` is the hook for the LOM Space to talk to LOM.  Since the only thing that should be stored in LOM is AIPs, I don't want to tell LOCKSS about any non-AIP (or at least, non-package) files that are accidentally put in LOM Space.  Also, it expects `package` to actually exist (and calls functions on it), so this is a check to ensure it does.

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -486,6 +501,497 @@ def post_move_from_storage_service(self, staging_path, destination_path, package
+        # to track information on it over time
+        if package is None:
+            return
+        # Post to Lockss-o-matic with the create resource atom entry
```
This implies something else is meant to be happening here, rather than just the logging entry?

Hwesta replied:
`Post to Lockss-o-matic with the create resource atom entry` is what the whole function does; this is intended as a section header rather than a description of the next line.  Is there a better way to designate this?
